### ISSUE TITLE
TF-Lite micro low performance

### ISSUE THREAD

**Issue by dimtass** [tensorflow/tensorflow#30953]
This issue regards the tflite micro.

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04.2
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: STM32F746NG
- TensorFlow installed from (source or binary): github
- TensorFlow version (use command below): 1.14.0-718503b075d
- Python version:
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I've ported a model from Keras to TF and trained it. Then I've converted the model to tflite.

One issue is that the tf.lite.Optimize.OPTIMIZE_FOR_SIZE doesn't work as the tflite-micro complains that one of the layers is not compatible with int8 types. Anyway, I've converted the model without quantization and also used CMSIS-NN. The performance is extremely slow. It's 40x times slower on the same CPU compared to X-CUBE-AI API. 

**Describe the expected behavior**
I expected the inference to be much faster especially for the depthwise_conv.cc kernel which supports the cmsis-nn. It seems that using cmsis-nn, makes an insignificant difference.

**Code to reproduce the issue**
I have a repo here that you can use for validation:
https://bitbucket.org/dimtass/stm32f746-tflite-micro-mnist

**Other info / logs**
Also, I've written a blog post with the issue here:
https://www.stupid-projects.com/machine-learning-on-embedded-part-3/

**Edit**: I've found out that I haven't enabled the FPU during the g++ compilation. With the correct architecture flags (-mthumb -mcpu=cortex-m7 -mfpu=fpv5-sp-d16 -mfloat-abi=hard), the performance is now 3x times faster, but it's still ~12x times slower than the x-cube-ai.

This is the execution time for each layer with comparison to x-cube-ai. Have in mind that the x-cube-ai merges some layers, so I'm using '-' and the sum in the next layer:

Layer | tflite-micro /soft-float (msec) | tflite-micro /hard-float (msec) | x-cube-ai (msec)
-|-|-|-
DEPTHWISE_CONV_2D | 235 | 69 | -
MAX_POOL_2D | 23 | 7 | 11.2
CONV_2D | 2346 | 733 | -
MAX_POOL_2D | 7 | 2 | 57.19
CONV_2D | 348 | 108 | 8.69
FULLY_CONNECTED | 5 | 3 | -
FULLY_CONNECTED | 0 | 0 | -
SOFTMAX | 0 | 0 | 2
TOTAL TIME | 2964 | 922 | 78.2

That's almost 40 times slower on the same MCU. I'm not sure if I'm doing something so terribly wrong or this is the real performance of the API.

I really hope for some input. Thanks!

**Comment by dimtass** [comment]
I'll also edit the main post, but I've found a stupid mistake that I've made.
I didn't enabled the FPU on the STM32F746. Now that I've enabled it, it's 3x times faster than before, so the inference time dropped from 2964 to 922ms.

But it's still much slower that the x-cube-ai implementation. It's now ~13x times slower, which is still too much.

I've also pushed the changes to the repo.

**Comment by Raukk** [comment]
I'd like to add my 2 cent's here;
TLDR; I think TF-Lite-Micro is using 64-bit Floats when it should be using 32-bit Floats. 


Based on the information provided by Dimitris [dimtass] in his articles, when using TF-Lite-Micro each 'Multiply–accumulate operation' (or equivalent) is taking approximately 70 cycles even with the (32-bit) FPU enabled (and approximately 225 cycles without FPU). To me, that seems way too slow for a 32-bit Cortex-M7 with a 32-bit FPU, even if you assume overhead or poor optimization on things like memory access, casting, etc. The fact that going from No FPU to using the FPU only gave a 3x speed up, makes me question what's happening here; specifically, if its working correctly, then a (32-bit) float multiply and addition on the FPU should be way under 70 cycles (even if it's not using the specialized Multiply–accumulate).  

Additionally, the fact that the X-Cube-AI framework runs the exact same network (same TF-Lite model and file) on the exact same hardware, and is within the expected performance range says something is very wrong with the TF-Lite run. 
Note: The X-Cube-AI averages 5.5 cycles per Multiply–accumulate operation (averaged over a whole prediction), which is 12 times faster than TF-Lite. 

The simplest explanation is that Double precision Floats (64-bit) are used when it should be (32-bit) Floats. This would force the usage of Software 64-bit Floats because the FPU is only 32-bit, the FPU would speed up some of the individual operations, but still be an order of magnitude slower than the 32-bit Float ran raw on the FPU as it should be. 


Note: I have not ran this myself, I am making inferences from the numbers and data provided by Dimitris in his Articles.
Note: I'm not a Micro Controller guy, if I made a stupid mistake, or misunderstanding, then let me know so I can learn.
Thanks,

**Comment by dimtass** [comment]
Today morning I thought why not build tflite-micro on my workstation and test it with the python implementation. I thought that since my workstation doesn't have any GPU acceleration and only the CPU is used, then I should get quite similar results.

At this point I need to mention that I'm not aware of the internal and the implementation on either C++ or python API, but my guess is that since only the CPU is available, then they should perform about the same.

I've created this repo here:
https://bitbucket.org/dimtass/tflite-micro-python-comparison

I'm explaining in the README with more detail how to build and run the tests. Don't mind about the cmake files that may refer stm32 at some points as I've just stripped the cmake project I've mention in the first post. So it's the same code with [this repo](https://bitbucket.org/dimtass/stm32f746-tflite-micro-mnist).

As I'm explaining in the readme I'm using the exact same model and input data.

My results for a 1000 inference run are:

tflite-micro | python
-|-
2.223457 ms | 0.199918 ms

So the python implementation is 11.2 times faster than the C++ code.

Although these are different APIs, I would expect that the performance should be same, as only the CPU is involved. **Unless**, the python API spawns multiple threads in multiple cores. That would explain the difference.

I've tried this command to run the python benchmark on a single cpu and I got the same result as without it, but tbh I don't know if that command works with python and actually prevents other threads to be spawned to other cores.
```sh
taskset --cpu-list 1 python3 python_benchmark.py
```

**Comment by Raukk** [comment]
Well, that would disprove my theory, since your workstation probably has a 64-bit FPU.

Can you run 10,000 or 100,000 inference runs to get a bigger metric?
A 2ms difference is hard to be sure it's not some other limitation (like HDD, OS, etc). 
Or is the time in ms mentioned (2.223457 ms, 0.199918 ms) the average per inference over the 1000 inferences? (making the total over 2 seconds for tflite-micro)

**Comment by dimtass** [comment]
Hi Raukk, yes the result is the average of the 1000 inferences. I've tried more and it seems a constant difference. For the tflite-micro running on the CPU the average time is 2ms, for python is 200us.

I'm afraid that this might be because the python backend is using all the cores/threads. I'm not sure if this `taskset --cpu-list` can actually stop python to spawn other threads to other cores.

Next, I want to try with some gcc flags and optimizations for my CPU. Maybe the python backend is using binaries with those optimizations enabled.

If I have time I'll try the same test on various SBCs I have around like a RPi3+, nanopi-neo4 and a couple of others. After this week I'll be off for 2 months and I won't have access to my equipment to run further tests :/

**Comment by Raukk** [comment]
I ran the Python script version (both 1,000 and 100,000 iterations) on an old laptop (Windows, i5) and the average time was 0.55ms on average. CPU was i5-3317U, with clock at 1700 MHz with single core-boost to 2600 MHz or both core 2400 MHz.

When running the Resource Manager shows all 4 CPUs (2 Core with Hyper threading) under heavy load, and CPU frequency at 140% which matches 2400 MHz. 
It reports the python usage as only about %50 but system usage at 80%. 
I'd take those values with a grain of salt. 
  
I then ran it on my Ryzen 7 (8 cores/16 threads), where CPU 8 got pegged, a few others showed minor load, and the rest were flat lined. Python usage was reported at ~10% with is more than 16th.
The Ryzen averaged 0.35ms at a sedate ~3.0 GHz, so the speed seems to be tightly coupled to the clock rate.
My guess would be that standard tensorflow was respecting the single threadedness for the actual math and calculations but was probably also doing everything else in a second thread (data loading, overhead, etc), just a guess.

While the fact it's probably running ~dual threaded may give it some advantage, I'd say it doesn't look like it gave more than double. Which would still leave at least a 5x slowdown. 

I didn't run the TF-Lite-Micro one yet, so, this is only for the python script you gave.

**Comment by dimtass** [comment]
I've just verified that the python backend spawns several threads in the background even if it's limited in one cpu, therefore my test is meaningless...

[python screeshot](https://ibb.co/88CNkRc)
[tflite-micro screenshot](https://ibb.co/2Ykd7CJ)

**Comment by Raukk** [comment]
@dimtass I wouldn't say it's "meaningless," it just means that you can't draw a 1-to-1 comparison.
Specifically, I don't believe that would result in a 10x speedup, since your report shows Python using 100%+30%+30%. Even with considering async-loading, pre-processing, or other things that the secondary thread could be doing, I cant see 30%+30% resulting in a 10x performance increase.
If the CPU it runs on is true single core without hyper threading, then the comparison should be accurate, I might try sticking it in a VM that only has one core, if I have time.


@petewarden Has anyone one the Tensorflow-Lite team looked at this (or [Read Dimitris Article](https://www.stupid-projects.com/machine-learning-on-embedded-part-3/) ) ?
I am not a TF-Lite Developer (and never will be), but I am interested in the results of Dimitris experiments, and the performance for TF-Lite looks to be very-very slow for a framework designed around the principles of running efficiently on low end hardware.

**Comment by dimtass** [comment]
I've just added more benchmarks for tflite-micro on various SBCs [here](https://www.stupid-projects.com/benchmarking-tensorflow-lite-for-microcontrollers-on-linux-sbcs/).

I wish there was a #define switch to enable threads to the tflite-micro API :)

**Comment by Raukk** [comment]
It's been 3 months, has anyone from the tensor flow project even looked at this?

**Comment by njeffrie** [comment]
Apologies for the delay and thanks for your feedback.  I read the article and you made a lot of good points for us to improve on.  Recently we have been digging into performance optimization, and working with ARM to land CMSIS-NN optimized kernels for key ops (conv, depthwise_conv, fully_connected).

The version of TFLite Micro you were using had uint8 CMSIS-NN optimizations crafted specifically for our speech example, so you would not have seen any speed-up for other models.  We recently received general-purpose CMSIS-NN kernels for int8 quantized models, which should speed things up significantly.  We realize there is still a long way to go to reach the performance numbers you saw using X-CUBE-AI. Our eventual goal is to get as close as possible to (or exceed where possible) platform-specific optimized libraries like X-CUBE-AI while providing a portable platform. We have also recently been working to improve and consolidate our documentation.

We realize there is a long way to go, so thank you for the hard work and clear feedback.

**Comment by freddan80** [comment]
Hello @dimtass ! Let me know what performance numbers you get with CMSIS-NN using int8 quantization. X-CUBE-AI uses CMSIS-NN kernels as well, so we expect a similar performance. Thanks for showing interest in CMSIS lib!

**Comment by dimtass** [comment]
Hi @freddan80 there's been quite some time that I haven't follow up the recent changes in tensorflow. Currently, I'm quite busy to setup everything again, but I plan to do so at some point in the near future in order to evaluate TF again. Therefore, I'll update this thread with my findings when it's done. Thanks!

**Comment by dimtass** [comment]
Hi @freddan80, I've run another round of benchmarks after I've ported the current stable 2.1.0 version. I've updated the repo [here](https://github.com/dimtass/stm32f746-tflite-micro-mnist). Before list the performance results, these are some comments regarding the new version:

- More kernels are supported for cmsis-nn, which is great
- I don't get any errors now when I load the optimized model, **but the firmware hangs**, so I'm not able to run any benchmarks.
- The tflite-micro performance of the 2.1.0 seems to be slightly worse than 1.4.0

This table shows the figures I get when running the uncompressed model @ 288MHz on the STM32F746

Kernel | 1.14.0 | 2.1.0 | diff (msec)
-- | -- | -- | --
DEPTHWISE_CONV_2D | 52 | 55.05 | -3.05
MAX_POOL_2D | 5 | 5.2 | -0.2
CONV_2D | 550 | 576.54 | -26.54
MAX_POOL_2D | 2 | 1.53 | 0.47
CONV_2D | 81 | 84.78 | -3.78
FULLY_CONNECTED | 2 | 2.27 | -0.27
FULLY_CONNECTED | 0 | 0.04 | -0.04
SOFTMAX | 0 | 0.02 | -0.02
TOTAL TIME: | 692 | 725.43 | -33.43

There are more detailed info in this post [here](https://www.stupid-projects.com/tensorflow-2-1-0-for-microcontrollers-benchmarks-on-stm32f746/).

Therefore, my main issue is that the optimized model although now it doesn't complain for INT8 the inference hangs, so I cannot get any results. The way I do the post training optimized is in this Jupyter notebook [here](https://github.com/dimtass/stm32f746-tflite-micro-mnist/blob/master/jupyter_notebook/MNIST-TensorFlow.ipynb) in cell 13. This is the code I'm using:

```py
tflite_mnist_model = 'mnist.tflite'
converter = tf.lite.TFLiteConverter.from_keras_model_file('mnist-tflite.h5')
converter.optimizations = [tf.lite.Optimize.OPTIMIZE_FOR_SIZE]
tflite_model = converter.convert()
flatbuffer_size = open(tflite_mnist_model, "wb").write(tflite_model)
print('The size of the converted flatbuffer is: %d bytes' % flatbuffer_size)
```

This will create a flatbuffer model which is 99KB in size, but as I've said this will hang the CPU. No further debugging was done from my side for this to verify the reasons.

What I've verified is that the cmsis-nn kernels seem to be used as supposed to, but the performance seems to be a bit worse than version 1.14.0.

Therefore, my questions are:
- what am I doing wrong regarding the post-optimization?
- Is there an extra configuration build flag that I need to enable?
- Any other proposals to debug the issue this issue?

**Comment by dimtass** [comment]
I didn't want to edit the above post and remove information, so I'm writing this one here with an update.

I've managed to make the optimized model load and it seems that from the `tensorflow/lite/core/api/flatbuffer_conversions.cc` I get that the tensor types inside the model are of type `kTfLiteFloat16` and `TensorType_INT8`, therefore the conversion seems to be working and also the model inference is running.

I've tested with both `-mfp16-format=ieee` and `-mfp16-format=alternative` flags during compile and it seems that I don't get any better performance, even with the model's weight optimizations.

These are the new results:

Layer | 1.14.0 | 2.1.0 | msec diff
-- | -- | -- | --
DEPTHWISE_CONV_2D | 52 | 55.03 | -3.03
MAX_POOL_2D | 5 | 5.2 | -0.2
CONV_2D | 550 | 576.44 | -26.4400000000001
MAX_POOL_2D | 2 | 1.51 | 0.49
CONV_2D | 81 | 84.79 | -3.79000000000001
FULLY_CONNECTED | 2 | 2.27 | -0.27
FULLY_CONNECTED | 0 | 0.04 | -0.04
SOFTMAX | 0 | 0.01 | -0.01
TOTAL TIME= | 692 | 725.29 | -33.29

Therefore, even with `TensorType_INT8` the performance is slow.

Any suggestions?

**Comment by freddan80** [comment]
Hello @dimtass! I think the most important issue to sort out here is the quantization of the model. It looks like the data type is float, which will not give you any performance.

To get the most out of TFLu and CMSIS-NN, please use int8 data type. Do you have a .tflite file?

A good place to check that the right function is called is for example here:

https://github.com/tensorflow/tensorflow/blob/64219fd7dabcc7f5b23cd72d0608378aaabb27ed/tensorflow/lite/micro/kernels/cmsis-nn/depthwise_conv.cc#L201

If you print `input->type` here:
https://github.com/tensorflow/tensorflow/blob/64219fd7dabcc7f5b23cd72d0608378aaabb27ed/tensorflow/lite/micro/kernels/cmsis-nn/depthwise_conv.cc#L341

You'll get the quantization type.

Cheers!

**Comment by dimtass** [comment]
Hi @freddan80 and thanks for the reply. I've used keras and tensorflow to build the model (you cen find the jupyter notebook [here](https://github.com/dimtass/stm32f746-tflite-micro-mnist/blob/master/jupyter_notebook/MNIST-TensorFlow.ipynb). Also, [this is](https://github.com/dimtass/stm32f746-tflite-micro-mnist/blob/master/jupyter_notebook/mnist-tflite.h5) the keras model that I've used for post-quantization.

The script that I've used for the quantization is taken from the tensorflow online documentation [here](https://www.tensorflow.org/lite/convert/quantization) and its the following:

```py
tflite_mnist_model = 'mnist.tflite'
converter = tf.lite.TFLiteConverter.from_keras_model_file('mnist-tflite.h5')
converter.optimizations = [tf.lite.Optimize.OPTIMIZE_FOR_SIZE]
tflite_model = converter.convert()
flatbuffer_size = open(tflite_mnist_model, "wb").write(tflite_model)
```

I've also added a printf in the line you suggested and indeed is seems that the type I get from the compressed model is `kTfLiteFloat32` while I was expecting `kTfLiteInt8`. That means that there's something going on with the conversion.

I've checked that the *.tflite I get after the conversion is much smaller (99.3KB) compared to the one without without the conversion (367KB), therefore I guess that it is processed. Although the fact that the model is much smaller and the input type is wrong and the non-sense results I get from this model when I run the inference, it means that there's something wrong with the conversion.

Do you have any suggestions on that? Is there any other way or tool to try to optimize an *.h5 model?

Thanks!

**Comment by freddan80** [comment]
@dimtass Sorry, I don't have much experience with model conversion. Perhaps @petewarden can point to someone.

Cheers!

**Comment by lheim** [comment]
Hey @dimtass, according to your code snippet and your Jupyter notebook you only quantize the weights - the inference computation is done in float32. Therefore you see no significant acceleration using `cmsis-nn`. See more [here](https://www.tensorflow.org/lite/performance/post_training_integer_quant).
> At inference, weights are converted from 8-bits of precision to floating point and computed using floating-point kernels. This conversion is done once and cached to reduce latency.

The following snippet provides a full int8 quantization (including computation). However, there is currently an issue with int8 input and output see #38285 

```python
def representative_dataset():
  for x in x_test_normalized:
    yield [np.array([x], dtype=np.float32)]

converter = tf.lite.TFLiteConverter.from_keras_model(my_model)
# Set the optimization flag.
converter.optimizations = [tf.lite.Optimize.DEFAULT]
# Enforce full-int8 quantization (except inputs/outputs which are always float)
# So to ensure that the converted model is fully quantized 
# (make the converter throw an error if it encounters an operation it cannot quantize), 
# and to use integers for the model's input and output, 
# you need to convert the model again using these additional configurations:
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter.inference_input_type = tf.int8  # or tf.uint8
converter.inference_output_type = tf.int8  # or tf.uint8
# Provide a representative dataset to ensure we quantize correctly.
converter.representative_dataset = representative_dataset
tflite_model = converter.convert()
open('./TFLite-model/LeNET-MNIST_int8ops.tflite', 'wb').write(tflite_model)
``` 


Try running your code on the STM32F7 again with the updated model. You should see some significant inference time improvement. 

In my premilinary results I see a speedup of 10x compared to the tflite-model with only the weights quantized.

**Comment by dimtass** [comment]
Hi @lheim I've tried the above snippet and I'm getting an error that `x_test_normalized` is not set, which OK it makes sense because I haven't defined it anywhere. How did you created this normalized set? Thanks!

**Comment by lheim** [comment]
@dimtass there is no special requirement from the TF Lite converter side. 
You only need to provide a *representative* dataset which is the same format as your training samples. In my case I used my whole normalized test set, as my network was trained on normalized data. Usually less samples are also fine (~100).

**Comment by mohantym** [comment]
Hi @dimtass !
I agree with @lheim'c [comment ](https://github.com/tensorflow/tensorflow/issues/30953#issuecomment-633905619) on using [integer quantized ](https://www.tensorflow.org/lite/performance/post_training_integer_quant)models in microcontrollers for improving performance.

Normalization of Mnist model is the normalization process mentioned in Mnist documentation.

Inference can be further fastened through [on-device training](https://devpost.com/software/handwritten-digit-recognition-on-microcontroller). 


Thank you!
